### DISCUSSION TITLE
Experiences with Getting Started

### DISCUSSION THREAD

**Discussion by AljenU** [automaticanalysis/automaticanalysis#254] [General]
Please make sure to preferably use the "write a reply" on a specific comment, instead of creating new threads in the discussion.

Shortly, I will be working on #242. In order to do that, I had to get automaticanalysis working on my system. Since this is my first time working with AA, i tried to get set up using the Getting Started and the examples. I found that the combination did not include all the information that i needed. Therefore, i am sharing my experiences here, with some proposals on how to improve the new user experience.

-=-=-

What i want to see in a getting started guide: information on what steps i need to take, up to the point where i have run the software at least once successfully. 
 
For AA, this should include 
1. Getting the software 
2. Putting AA main folder on the matlab path 
3. Getting a minimal parameter xml 
4. Getting the demo dataset(s) 
5. Knowing which example to start with 
 
In my recent experience, only the " Getting the software" step was described such that I got through that step without incident. 

The " Putting AA main folder on the matlab path" step should be part of the Getting Started guide, under the installation heading. What needs to  be put on the path for a certain matlab toolbox to work differs per toolbox, so it should be made explicit that only the folder that contains aa_ver5.m needs to be put on the path, and that every script that wants to use aa, should start with calling aa_ver5 (as can also be seen in the examples).

-=-=- 

What was needed for 3, 4 and 5 was less clear. 

**On the parameter xmls:**  

- The Getting started guide does talk about Parametersets, but it does not explicitly state how i should go about getting set up with a first parameter xml, and where to put it / them. 
- If i start to try out an example without doing anything for a parameter xml, then (in the call to aa_recipe) i am prompted whether i want to create one, and indeed a aap_parameters_user.xml is created. However, the xml that is created does not allow me to continue the example without errors.
- **Proposed**: Extend what is now in aa_recipe/create_minimalXML to actually create a minimal viable xml. (Details later) 

**On the demo datasets:** 
 - The aa_downloaddemo works nicely, but only if rawdatadir in the parameter xml is set up in a specific way as expected by the example that i am running. 
 - Some examples expect the data to have been downloaded and the user to provide extra info from the m script, while others use downloaddemo and the aap_parameters_user.xml. 
 - **Proposed**: Have a dedicated example script to download the demo datasets (Details later) 

**On the examples:** 
 - There is no uniform naming of the examples 
 - Example scripts and utility function files are mixed in one folder 
 - Some of the example scripts are not mentioned in the README in the examples folder 
 - I have this parameter xml file, where i can set parameters, yet then the scripts are also settings parameters in the aap structure. And even setting parameters that for other example scripts i needed to set explicitly in the parameter xml. 
 - Even when i do have the parameter xml correctly and the demo datasets downloaded, the examples do not work out-of-the-box  (Details later) 
 - **Proposed**: Make clear the difference between what is set in the parameter xml, and what is set in a run script. (Details later) 
 - **Proposed**: Have a dedicated run-this-one-first example, to introduce the user to the parameter xml and the aap structure. (Details later)

**Comment by AljenU** [comment]
**Proposed**: Extend what is now in aa_recipe/create_minimalXML to actually create a minimal viable xml.

Currently, only the seed parameterset is included, and acq_details/root is set. The resulting parameterset has two types of problems before it is a viable parameterset for use with the examples. This assumes that aa_parametersets/aap_parameters_defaults.xml was selected as the seed. The problems are all related to directory_conventions
1. Some required paths for the examples have not been adapted for the specific computer, mainly rawdatadir and the location of spm
2. Various other paths refer to locations that do not exist on the specific computer (rawmeegdatadir, fsldir, freesurferdir, condorwrapper, ROIdir, realtime/path)

To take into account:
- Adapt aa_recipe/create_minimalXML, or does that function have a use in other situations such that it cannot be dedicated to this proposal?
- Assume the current state of aa_parametersets/aap_parameters_defaults.xml as input to the function?
- Creating a minimal viable xml would be (only?) used once in a run-this-one-first example

Alternatives / additions to main proposal:
- Adapt aa_parametersets/aap_parameters_defaults.xml so all the mentioned paths are empty instead of containing (possibly) non-viable paths
- Add a second ???_defaults.xml that has these paths empty

**Reply by AljenU** [reply]
Preferably, already take into account that $HOME does not work on Windows.

**Reply by AljenU** [reply]
How much does this need to be backwards compatible? And, how much user input can be asked?

In any case, i would want to adapt helloworld and aarecipe, such that a dedicated first step in the first tutorial is to explicitly set up the user parameter xml (that will still derive from the defaults.xml).
In that step, the required directories should all be set correctly. For the not-required-for-tutorials directories, the user should be notified somehow that if he does have the tools installed, the path can be set. 
If the user selects a defaults_SITESPECIFIC, i would assume that no changes are needed except the acq_details/root that is already asked for. If the general defaults is selected, also rawdatadir and the location of spm need to be specified.

My personal preference would be to adapt the aap_parameters_defaults.xml so it assumes nothing about the directories on the users computer, i.e. that all settings that need to be 'absolute' directories are empty.
However, this is not backward compatible for users that use the git repo directly, and in addition use a parameter xml that derives from aap_parameters_defaults, and rely on one of those directory defaults. Then again, how many users would that be? And would those not mainly be developers?

An alternative would be to either ask about all of them explicitly during the creation of the user xml, like is done for acq_details/root. However, that would add some 6 extra questions / dialogs, which is not very user friendly.
Second alternative, to set them explicitly empty in the user xml, thus overriding the value from the defaults.xml. The effect then will be similar to adapting the defaults.xml itself.

**Reply by tiborauer** [reply]
We have already talked about the defaults.xml with @jones-michael-s, and agreed that the most transparent and least confusing option if all absolute path settings are empty.

**Reply by AljenU** [reply]
I created #297 for making the default xml emptier.

That leaves only "Some required paths for the examples have not been adapted for the specific computer, mainly rawdatadir and the location of spm" as what needs to be adapted in setting up a user parameter xml, with "In any case, i would want to adapt helloworld and aarecipe, such that a dedicated first step in the first tutorial is to explicitly set up the user parameter xml (that will still derive from the defaults.xml)."

**Reply by AljenU** [reply]
And now there is issue #301 for the other adaptations to helloworld and aarecipe to create a minimal-complete user parameter xml.

**Comment by AljenU** [comment]
**Proposed**: Have a dedicated example script to download the demo datasets

Already the downloaddemo and how it is used, has assumptions on where the datasets will be. And, only two datasets are used in all the examples. And, the downloaddemo requires a specific rawdatadir entry.

Thus, a dedicated example script that explains the interaction of the demodatasets and the rawdatadir entry in the parameter xml.
And adaptations to the downloaddemo function so you are choosing which one out of only two options you want to use. 
And then adapt all examples to use the function, and the rawdatadir entry in the parameter xml (instead of some of them adapting that setting within the script).

To take into account:
- This can be one of the parts of a run-this-one-first example

**Reply by jones-michael-s** [reply]
I will never be convinced that any kind of automatic or scripted data download is a good idea. Is it really so offensive to expect an aa user to be able to locate and manually download a dataset? 

However, I appear to be in the minority on this point and I simply don't have the strength to fight about it anymore.

**Reply by tiborauer** [reply]
I hear you, Mike, and you are right when considering developers. That is why I think you should be able to opt out of the automatic download.
A typical user, on the other hand, may have a dataset only for their own research and may not want to bother with locating and retrieving the example data. For them, the automatic download might be useful.

**Reply by AljenU** [reply]
First step could be to just adapt downloaddemo so it can be used in the tutorial scripts without requiring any user adaptations of the scripts.
For this, adapt so it is dedicated to downloading only specific sets (have less inputs to the script).
And (not sure if this is already approximately there) if there is only one rawdatadir, and it does not have aa_demo part, create such a folder under the root rawdatadir.

**Comment by AljenU** [comment]
**Even when i do have the parameter xml correctly and the demo datasets downloaded, the examples do not work out-of-the-box**

TODO: Create an issue for "Make as many AA examples work with just SPM as possible", when the various problems listed here have been assessed for a correct solution.

1. Most examples can run with just AA and SPM, except that the presence of FSL is implicit in some places.
   - In aamod_segment8/diag, a T1 file is read from aap.directory_conventions.fsldir. This gives an error when fsl is not present. The same file is hardcoded in 5 other places in the AA code! In one of those places (aamod_freesurfer_register), also an SPM provided alternative is given. Applying that alternative in aamod_segment8 allows the examples that use segment8 to run to completion. **Proposed**: put the hardcoded filename in one place, a getter function that also has something like a allow_alternative input option. And use that function in all 5 places, with the allow_alternative option set depending on where the function is called from.
   - In aa_provenance, the fls version file is read, without checking if it exists. This gives an error when fsl is not present. Adding a check on the presence of the fsldir and the version file prevents the error. **Proposed**: add this check.
2. aas_processBIDS fails on filepaths that have a . somewhere in it.
   - In particular, in two places in aas_processBIDS, an expression like `[~,n,~]` = fileparts(strtok(cfname{1},'.'));` is used. This does not work correctly for stripping a file extension, when there is a . somewhere in the path name. **Proposed**: probably a double fileparts can be used, have a quick discussion on that.
   - In general, strtok is dangerous when processing paths and filenames. The default "split on whitespace" can happen to be within a path or filename, not just in between when processing a list of them. This also holds for custom split tokens that are an allowed directory or filename component. **Proposed**: Add an issue to check whether the 127 calls to strtok are vulnerable to this problem.
   - Even more general, strtok is dangerous on anything that comes from user input, unless that input has been thoroughly checked. **Proposed**: Add an issue to check whether the 127 calls to strtok are vulnerable to this problem. And require that each use of strtok has a comment that documents why it is safe to use, when that is not immediately clear from the surrounding code.
3. Module capitalization in aa_example_helloworld_model
   - The module aamod_biascorrect_meanEPI is referenced, whereas the actual xml file in aa_modules is aamod_biascorrect_meanepi.xml. This gives an error. **Proposed**: Fix capitalization.
4. Default wheretoprocess in bids_ds000114_aa
   - The default is batch, which failed for me, details to be discussed. Using matlab_pct or parpool did work correctly out-of-the-box. **Proposed**: use on of these other options as default in the example. And, have an example specifically on cluster setup? **EDIT** For bids_ds000114_aa, this example should be marked as advanced, as it uses certain freesurfer functions. On batch, see nee next item.
5. Using 'batch' as wheretoprocess, in R2021b
   - The batch option uses the reqpath. The cellindex way used to compose the reqpath, will fail to add the 'main' folder of a certain toolbox, if the folder is specified with a trailing pathsep character (i.e. a / on linux). Thus the batch jobs can error on failing to find certain functions, as they are not on the path for the tasks matlab process. **Proposed** Fix reqpath creation to take possible trailing pathseps into account.
   - The batch option uses aas_getTaskDuration, which uses an matlab class and function of the parallel.internal namespace. It looks like the function has changed (in R2021b or earlier), and it now gives an error. **Proposed** Do not use the method from the parallel.internal class. Even, do not use the parallel.internal class at all then, since it was only introduced for this specific method only.

**Reply by jones-michael-s** [reply]
> in aamod_segment8/diag, a T1 file is read from aap.directory_conventions.fsldir. This gives an error when fsl is not present.

Yep. This is a structural template file and I never understood why aa uses the one from FSL. Is there not an alternative in SPM? Or could we not simply include this one file as part of the distribution? 

To add insult to injury, diag() isn't even necessary for processing. It just creates diagnostic images, which are actually quite nice but I always thought it was still a bit silly to require the user to install FSL just for this.

> Module capitalization in aa_example_helloworld_model

That's my fault. I wrote the example for a previous aa version and I thought I caught all of the name changes. Your intuition is correct -- the fix is just a simple name change to aamod_biascorrect_meanepi. I'll include this fix in my next PR.

> Default wheretoprocess in bids_ds000114_aa

I'm a less concerned about this. Example scripts are just that: examples. They're not intended to run as is (the PR test scripts are, but that's a different issue). However, this specific glitch does speak to the larger issue that the aa documentation is sketchy. I think I'll be taking point on that in the coming months, other time constraints permitting.

**Reply by tiborauer** [reply]
On the note on the segment8/diag. There is, indeed, a T1 template in SPM, however, it is very smooth and you can barely notice the structure. That is why I decided to use FSL's instead. 
However, I understand that requiring FSL just for diag is not an inclusive approach. We may change the code to use SPM's if FSL's is not available.

**Reply by jones-michael-s** [reply]
Could we simply include a copy of MNI152_T1_1mm.nii.gz in the distribution? I don't think diag uses any other FSL functionality.

**Reply by tiborauer** [reply]
I would avoid including any template. However, we have a setting `aap.directory_conventions.templatedir`, which could be used for that Alternatively (and more unambiguously), we could use `aap.directory_conventions.T1template` to point to any template. And wherever the SPM template is needed (e.g. by SPM registration functions), we make it explicit using `spm('Dir')`

**Reply by AljenU** [reply]
I've edited the comment for item 4 and added an item 5

4. Default wheretoprocess in bids_ds000114_aa

    The default is batch, which failed for me, details to be discussed. Using matlab_pct or parpool did work correctly out-of-the-box. Proposed: use on of these other options as default in the example. And, have an example specifically on cluster setup? EDIT For bids_ds000114_aa, this example should be marked as advanced, as it uses certain freesurfer functions. On batch, see nee next item.

5. Using 'batch' as wheretoprocess, in R2021b

    The batch option uses the reqpath. The cellindex way used to compose the reqpath, will fail to add the 'main' folder of a certain toolbox, if the folder is specified with a trailing pathsep character (i.e. a / on linux). Thus the batch jobs can error on failing to find certain functions, as they are not on the path for the tasks matlab process. Proposed Fix reqpath creation to take possible trailing pathseps into account.
    The batch option uses aas_getTaskDuration, which uses an matlab class and function of the parallel.internal namespace. It looks like the function has changed (in R2021b or earlier), and it now gives an error. Proposed Do not use the method from the parallel.internal class. Even, do not use the parallel.internal class at all then, since it was only introduced for this specific method only.

**Reply by tiborauer** [reply]
I think, if batch is fixed, then it can stay as processing queue for bids_ds000114_aa (which is an advanced demonstration, for sure).

**Comment by AljenU** [comment]
**Proposed**: Make clear the difference between what is set in the parameter xml, and what is set in a run script.

Actually, this difference is not clear to me. I would like to understand better what is the recommended way. Please do provide comments.

**Reply by jones-michael-s** [reply]
I agree this is confusing. It's usually just a matter of user preference (although there are some options that are awkward if not impossible to specify in the xml). 

Again, better documentation is the fix.

**Reply by tiborauer** [reply]
See my [answer below](https://github.com/automaticanalysis/automaticanalysis/discussions/254#discussioncomment-1502846).

**Comment by AljenU** [comment]
**Proposed**: Have a dedicated run-this-one-first example, to introduce the user to the parameter xml and the aap structure.

The helloworld example appears to have this intention. However, it does not explain about the parameter xml and the aap structure. And it wants the user to set certain variables within the script, whereas other example scripts assume that the parameter xml has all the information that is needed.

**Reply by jones-michael-s** [reply]
As in the choice between userscript (i.e., the mfile) and tasklist (i.e., the xml), some users may prefer to provide settings in the parameter file, and others may prefer to set parameters in the script (some parameters MUST be set in the parameter file for PR testing, but that is a different issue).

Again, better documentation is the fix.

**Reply by tiborauer** [reply]
I would rather distinguish three sources:
- parameter set (XML)
- tasklist (XML)
- user script (M-file)
This is typically the order of variability (low to high), as well.

A parameter set is designed to store information about the environment, which is usually invariant for an institute. 
Tasklist describes the structure of the project. Some parameters can be set here, as well; however, I prefer to keep them in the user script for the sake of transparency.
The user script is about customising the analysis and overwriting any default setting. 

Theoretically, the whole analysis can be specified just from within the user script by creating and editing the `aap` structure. However, it would be very inconvenient and lead to a lot of redundancy.

**Comment by jones-michael-s** [comment]
It occurs to me that we currently have two types of example scripts in /examples, not all of the which are intended to run as-is (the testing scripts in /developer are intended for efficient PR testing and are a separate thing).

This leaves a new user at a loss where to begin. Perhaps we should divide the example scripts into -- for lack of better terms -- -"demonstration" and "FYI" (or maybe "beginner" and "intermediate"?) The demonstration scripts would be things like helloworld that a new user really are intended to teach how to use aa. The FYI scripts are the older scripts that give examples of more advanced functionality that the intermediate user might incorporate into their own work but aren't necessarily where a new user would begin.

Certainly the current naming jumble in /examples doesn't help and I'm as much to blame as anyone. I recently added all the `aa_example_*` scripts, which maybe should really be named `aa_demo_*.` A new user really should be able to run these. Then, all the `demo_*` scripts should be renamed `aa_example_* `(not sure what to do with` bids_ds000114_aa` and `meg_aa` and the cbu stuff)

Anyway, just thinking out loud...

**Reply by tiborauer** [reply]
It is a great suggestion, however, I suggest another terminology:

- `gettingstarted`,`starter`, or `start` could indicate scripts to start with as a newbie 
- `demo` to demonstrate a(n advanced) functionality

**Reply by AljenU** [reply]
I've added #256 to follow up on this.
